File: guests/python/scroll.py
"""The scroll conformance scene, Python port — the viewport grows so
the enclosing track constrains it (an unconstrained viewport hugs its
content and nothing overflows); the bottom button, reachable only by
scrolling, proves the scrolled-to content is live. See
guests/rust/scroll.rs and tools/scenes/scroll.steps."""

import sys

import kaya

app = kaya.App()


def bottom_clicked():
    status.set("bottom clicked")


with app.window(title="scroll"):
    status = kaya.signal("at top")
    with kaya.column():
        kaya.label(bind=status)  # label#0
        with kaya.scroll(grow=1):  # scroll#0
            with kaya.column():
                for i in range(1, 30):
                    kaya.label(bind=kaya.signal(f"row {i}"))
                kaya.button("bottom", on_click=bottom_clicked)  # button#0

sys.exit(app.run())

File: guests/js/scroll.ts
// The scroll conformance scene (tools/scenes/scroll.steps). The viewport
// GROWS: unconstrained it hugs its content and nothing overflows.

import * as kaya from "kaya-gui";

const app = new kaya.App();

function bottomClicked(): void {
  status.set("bottom clicked");
}

const { status } = app.window({ title: "scroll" }, () => {
  const status = kaya.signal("at top");
  kaya.column(() => {
    kaya.label({ bind: status }); // label#0
    kaya.scroll({ grow: 1 }, () => {
      // scroll#0
      kaya.column(() => {
        for (let i = 1; i < 30; i++) {
          kaya.label({ bind: kaya.signal(`row ${i}`) });
        }
        kaya.button("bottom", { onClick: bottomClicked }); // button#0
      });
    });
  });
  return { status };
});

app.run();
